source/common/network/listen_socket_impl.h

```diff
@@ -50,5 +50,55 @@ class UdsListenSocket : public ListenSocketImpl {
   UdsListenSocket(const std::string& uds_path);
 };
 
+class ConnectionSocketImpl : virtual public ConnectionSocket {
+public:
+  ConnectionSocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address,
+                       const Address::InstanceConstSharedPtr& remote_address)
+      : fd_(fd), local_address_(local_address), remote_address_(remote_address) {}
+  ~ConnectionSocketImpl() { close(); }
+
+  // Network::ConnectionSocket
+  const Address::InstanceConstSharedPtr& localAddress() const override { return local_address_; }
+  const Address::InstanceConstSharedPtr& remoteAddress() const override { return remote_address_; }
+  void setLocalAddress(const Address::InstanceConstSharedPtr& local_address,
```

Comment:
can we add some `ASSERT` here that verifies that someone does not violate the invariants that you set out in the function description?

Reply:
Good point. Done.